Set groundwork for sample caching and SampleBuffer rewrite

sakertooth opened the pull request:
This is my contribution to sample caching. Of course, this is just a draft and my idea as to how it should work, and it may differ from what someone else had in mind. Feel change to change and simplify what is necessary if you think it will work well.

This PR creates a seperate ``SampleBufferV2`` class and implements the ``SampleCache``. Hopefully, once ``SampleBufferV2`` is finalized, it can be reimplemented with the rest of the codebase with the required renaming. I did not want to change ``SampleBuffer`` just yet, as I belive there should be a new foundation first.

Below is an overview of the PR.

### `SampleBufferV2`
`SampleBufferV2` stores the raw sample data as immutable in memory. As such, it differs from `SampleBuffer` as its purpose is different. `SampleBufferV2` can then be shared using `std::shared_ptr` (which is why its immutable; nothing should be allowed to change it except itself for resampling). This class is not copyable and only movable, as copying this class will result in copying all the sample data, which can be fairly large.

### `Sample`
`Sample` is essentially the raw sample data but with its own characteristics (if it is meant to be played in reverse, for example). It acts as a "state" of `SampleBufferV2` in which the buffer *acts* differently but is not changed. This will most likely be the replacement for most of the use of `SampleBuffer` in the current codebase. For example, `SampleBuffer` can be replaced with `Sample` in `SampleClip` as the interface is mostly the same. However, it is just decoupled from the actual sample data, which is stored in `SampleBufferV2`.

### `SampleBufferCache`
When creating new sample buffers from either Base64 or a file, there needs to be a way to know if it has already been loaded into memory so it can be reused. `SampleBufferCache` can be used to cache and retrieve sample buffers without having to reload it back into memory.  It keeps track of sample buffers using ``std::weak_ptr`` and returning ``std::shared_ptr`` for those that want to use the sample buffer. The shared pointers returned have a custom deleter which deletes the sample buffer and removes the sample buffer from the cache when there are no more references to it.

### Other notes
* `HandleState` has been removed and has mostly merged with `Sample`.

###  TODO
* Add static open functions to `Sample`
* Add support for DrumSynth decoding in `SampleBufferV2`
* Implement `play` and `visualize` in `Sample

5/1/2022 - Progress. No visualization.
[Milestone 1](https://user-images.githubusercontent.com/20197645/166155493-150fcbfa-04c8-4649-b7eb-6407da529985.mp4)

sakertooth committed:
Rebase sample caching branch

sakertooth committed:
Remove HandleState

sakertooth committed:
Remove SampleCache in favor of SampleBufferCache

sakertooth committed:
Make SampleBufferV2 a QObject
Connect SampleBufferV2 to a sampleRateChanged slot
Add draft for Sample class

sakertooth committed:
Add support for base64 encoding/decoding

sakertooth committed:
Alter implementation for base64 encoding/decoding

sakertooth committed:
Fix compiler error with Sample.h 'loopEnd'

sakertooth committed:
Add resampling

sakertooth committed:
Add implementation to Sample
* Add Sample to CMakeLists

sakertooth committed:
Add numFrames() function
Add const to sampleFrame* in constructor

sakertooth committed:
Rework SampleBufferCache implementation
Use std::weak_ptr to keep track of buffers instead

sakertooth committed:
Im tired, brush up on implementation

JGHFunRun commented:
TBH I think this stands better if you just open a PR straight to the LMMS GitHub, you did do all the work (and I thank you for it), at least if you can I'm not sure if GitHub is capable of that but I'm guessing it is

sakertooth committed:
Add play functionality
Add SRC_STATE for resampling when changing pitch
Remove userWaveSample (might add it back in)
General cleanup

sakertooth committed:
Use deleteLater() instead of delete

sakertooth committed:
Add original sample rate
Make numFrames only constructor explicit

sakertooth committed:
Integrate the Sample class with other classes
* SampleClip
* SamplePlayHandle
* SampleClipView
* SampleTrack

sakertooth committed:
Add sample rate functionality
Brush up on implementation

sakertooth commented:
> TBH I think this stands better if you just open a PR straight to the LMMS GitHub, you did do all the work (and I thank you for it), at least if you can I'm not sure if GitHub is capable of that but I'm guessing it is

I figured that since a PR was already there for sample caching, I would just work under it. But yeah, I am able to make my own.
